### PR TITLE
log(c) for Gaussian integral

### PR DESCRIPTION
**Pull request by elib20**
### **User description**
**Context:** Sometimes the Gaussian integral functions produced incorrect results because the c values being multiplied together consisted of very small and very large numbers.

**Description of the Change:** Uses logarithms of c values to ensure more stability. This required adding the option for joinAbc to return the logarithm of c directly. Tests are updated accordingly.

**Benefits:** Stable Gaussian integrals.

**Possible Drawbacks:** None as far as I know.

**Related GitHub Issues:**

___

### **PR Type**
Enhancement

___

### **Description**
- Improved numerical stability for Gaussian integrals using logarithms

- Added `return_log_c` parameter to `join_Abc` function

- Updated integration functions to handle log coefficients

- Added tests for highly displaced squeezed states normalization

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["join_Abc function"] -- "adds return_log_c parameter" --> B["log coefficient handling"]
  B -- "passes to" --> C["complex_gaussian_integral_1"]
  C -- "uses log arithmetic" --> D["stable computation"]
  D -- "enables" --> E["normalized displaced states"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>gaussian_integrals.py</strong><dd><code>Core Gaussian integral logarithmic coefficient implementation</code></dd></summary>
<hr>

mrmustard/physics/gaussian_integrals.py

<ul><li>Added <code>return_log_c</code> parameter to <code>join_Abc</code> function<br> <li> Modified coefficient computation to use logarithms<br> <li> Updated integration functions to accept log coefficients<br> <li> Changed function signatures to handle <code>log_c</code> instead of <code>c</code></ul>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/641/files#diff-a7b779ea109bae98fb3be0cb7ed5f6dba566d967f72666d24fbd01015f7962dc">+26/-22</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>polyexp_ansatz.py</strong><dd><code>Updated ansatz trace method for log coefficients</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

mrmustard/physics/ansatz/polyexp_ansatz.py

<ul><li>Modified <code>trace</code> method to compute log of coefficient<br> <li> Updated call to <code>complex_gaussian_integral_1</code> with log coefficient</ul>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/641/files#diff-7c01e9c60d2d4a82b42f83c1e79989674c8a3a4a215091a23cfbfb7b4feca1c9">+5/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_gaussian_integrals.py</strong><dd><code>Enhanced tests for logarithmic coefficient functionality</code>&nbsp; </dd></summary>
<hr>

tests/test_physics/test_gaussian_integrals.py

<ul><li>Added parametrized tests for <code>return_log_c</code> parameter<br> <li> Updated test calls to use log coefficients where needed<br> <li> Added assertions for both log and regular coefficient modes</ul>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/641/files#diff-f64960aa126fc9053f056b46e568e9a1cd22ccb908dbccad294f7d67a1fbbb1b">+29/-14</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_displaced_squeezed.py</strong><dd><code>Added normalization tests for displaced squeezed states</code>&nbsp; &nbsp; </dd></summary>
<hr>

tests/test_lab/test_states/test_displaced_squeezed.py

<ul><li>Added new test for probability normalization<br> <li> Tests highly displaced squeezed states with large parameters</ul>

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/641/files#diff-165ef1da79743134a0a62ca2a50fa2ac888a5c6178fc72b6829c0612d06e7aef">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_polyexp_ansatz.py</strong><dd><code>Updated ansatz test for log coefficient usage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_physics/test_ansatz/test_polyexp_ansatz.py

- Updated `test_trace` to use log coefficient in function call

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/641/files#diff-921ac6091261c146640c3b3df644134e8872df68078997207977cb7f4231f158">+3/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_b_to_q.py</strong><dd><code>Updated BtoQ tests for log coefficient parameter</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

tests/test_lab/test_circuit_components_utils/test_b_to_q.py

- Updated calls to `join_Abc` with `return_log_c=True` parameter

</details>

  </td>
  <td><a href="https://github.com/XanaduAI/MrMustard/pull/641/files#diff-034dfd11764f85c16e8f21392e2ce9181dc26f97aba39a463b5f3e4a7a453ca9">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

**Comment by aplund** [comment]
For c = 0, numpy will raise a warning, but gives the "correct" result, i.e. -inf.  Jax doesn't seem to.  Numpy needs to have the divide by zero warning suppressed in this case.

File: mrmustard/physics/gaussian_integrals.py [hunk 1]
```diff
@@ -370,11 +374,11 @@ def true_branch_complex_gaussian_integral_1(m, M, bM, det_M, c, D, R, bR):
...
 
-    c_factor = math.sqrt(math.cast((-1) ** m / det_M, "complex128")) * math.exp(
-        -0.5 * math.sum(bM * M_bM, axis=-1),
+    log_c_factor = math.log(math.sqrt(math.cast((-1) ** m / det_M, "complex128"))) + (
+        -0.5 * math.sum(bM * M_bM, axis=-1)
     )
-    c_reshaped = math.reshape(c_factor, batch_shape + (1,) * (len(c.shape[batch_dim:])))
-    c_post = c * c_reshaped
+    log_c_reshaped = math.reshape(log_c_factor, batch_shape + (1,) * (len(log_c.shape[batch_dim:])))
+    c_post = math.exp(log_c + log_c_reshaped)
```
Comment (heltluke) [inline mrmustard/physics/gaussian_integrals.py:381]: I'm somewhat surprised that this is necessary. Can you point to tests that fail if we _don't_ do this? That is, are there tests that were not simply "updated" but also used as part of test-driven development?

**Review by heltluke — DISMISSED** [review]
Provided the `math.log` and `math.exp` calls aren't too costly, I don't really see a problem with merging this. I'm just surprised it's necessary.

File: mrmustard/physics/gaussian_integrals.py (hunk 1 above)
Comment (elib20) [reply mrmustard/physics/gaussian_integrals.py:381]: It is a relatively edge case bug. The new test I added in test_displaced_squeezed.py covers the example that originally produced the bug for me.

Comment (heltluke) [reply mrmustard/physics/gaussian_integrals.py:381]: The
```python
    @pytest.mark.parametrize("alpha", [10 + 10j, 18 + 18j, 25 + 25j])
    @pytest.mark.parametrize("r", [0, 1, 2, 3])
    def test_probability(self, alpha, r):
        """Tests that highly displaced squeezed states are properly normalized."""
        state = DisplacedSqueezed(mode=0, alpha=alpha, r=r)
        assert math.allclose(state.probability, 1)
```
got it. You don't even need a `DisplacedSqueezed` state to do this though, right? It's the large displacement that is causing an issue, as it leads to a very small `c` (as it's `exp(-|alpha|^2 / 2)`)? That is, the issue appears to be about numbers that are too small to be represented by `complex128`, not necessarily about the product of a large and a small number.

Comment (heltluke) [reply mrmustard/physics/gaussian_integrals.py:381]: Somewhat has me wondering if there should be warnings about creating states that will have such issues. Already, the probability is `nan` for a state like `Coherent(mode=0, alpha=26.64)`.

Comment (elib20) [reply mrmustard/physics/gaussian_integrals.py:381]: I just moved the test to change from being a test of DisplacedSqueezed to a test of Coherent.

I agree this won't fix everything. There will still be a limit where things break, as you've found. Not sure what best solution is.

On this branch, `Coherent(mode=0, alpha=26.64)` works. Things start to break down as you go higher than `Coherent(0, 38.2).probability`. By `Coherent(0, 38.7).probability` I get a warning about an invalid encountered in log.

Comment (heltluke) [reply mrmustard/physics/gaussian_integrals.py:381]: In the future, we should probably work to raise that `.probability` is not possible to be calculated for such states at the time they're created instead of only when `.probability` is called, and potentially consider using A,b,log(c) instead of A,b,c throughout the codebase, for for now I think the changes hare are good.

**Review by heltluke — APPROVED** [review]
